src/context.rs

```diff
@@ -332,24 +341,57 @@ pub async fn plan_to_object_store(
                 {
                     result?;
                 } else {
-                    // TODO: the object_store crate doesn't support multi-part uploads / uploading a file
-                    // from a local path. This means we have to read the file back into memory in full.
-                    // https://github.com/influxdata/object_store_rs/issues/9
-                    //
-                    // Another implication is that we could just keep everything in memory (point ArrowWriter to a byte buffer,
-                    // call get_parquet_file_statistics on that, upload the file) and run the output routine for each partition
-                    // sequentially.
-
-                    let mut buf = Vec::new();
-
-                    let mut file = tokio::fs::File::open(&partition_file_path).await?;
-                    file.read_to_end(&mut buf).await?;
-                    let data = Bytes::from(buf);
-
-                    store
-                        .inner
-                        .put(&Path::from(object_storage_id.clone()), data)
-                        .await?;
+                    let file = AsyncFile::open(partition_file_path).await?;
+                    let mut reader =
+                        BufReader::with_capacity(PARTITION_FILE_BUFFER_SIZE, file);
+                    let mut part_buffer =
+                        BytesMut::with_capacity(PARTITION_FILE_MIN_PART_SIZE);
+
+                    let location = Path::from(object_storage_id.clone());
+                    let (multipart_id, mut writer) =
+                        store.inner.put_multipart(&location).await?;
+
+                    let error: std::io::Error;
+                    loop {
+                        match reader.read_buf(&mut part_buffer).await {
+                            Ok(size) => {
+                                if size == 0 && part_buffer.is_empty() {
+                                    // We've reached EOF and there are no pending writes to flush
+                                    // TODO: as per the docs size = 0 doesn't actually guarantee that we've reached EOF
```

Comment:
Is this still true? Weird, since there's no other way to signal EOF (unless it returns a special Err type)?

Reply:
Yes, this is still true. I added a comment with a potential workaround using [stream_len](https://doc.rust-lang.org/nightly/std/io/trait.Seek.html#method.stream_len) (`nightly-only experimental API` atm).

I've also tried to get the true file size from metadata beforehand and keep track of the total read bytes, but they never matched somehow.